mlir/include/Gradient/IR/GradientOps.td

```diff
@@ -176,4 +177,40 @@ def VJPOp : Gradient_Op<"vjp", [
     let hasVerifier = 1;
 }
 
+def VectorInitOp : Gradient_Op<"vector_init"> {
+    let summary = "Initialize a dynamically resizable parameter vector.";
+    let results = (outs ParameterVectorType:$vector);
+    let assemblyFormat = [{ attr-dict `:` type($vector) }];
+}
+
+def VectorPushOp : Gradient_Op<"vector_push",
+        [TypesMatchWith<"type of 'value' matches element type of 'vector'",
+                        "vector", "value",
+                        "$_self.cast<ParameterVectorType>().getElementType()">]> {
+    let summary = "Append an element to the end of a parameter vector.";
+    let arguments = (ins AnyType:$value, ParameterVectorType:$vector);
+    let assemblyFormat = [{ $value `,` $vector attr-dict `:` type($vector) }];
+}
+
+def VectorSizeOp : Gradient_Op<"vector_size"> {
+    let summary = "Get the size of a parameter vector.";
+    let arguments = (ins ParameterVectorType:$vector);
+    let results = (outs Index:$size);
+    let assemblyFormat = [{ $vector attr-dict `:` type($vector) }];
+}
+
+def VectorLoadDataOp : Gradient_Op<"vector_load_data"> {
+    let summary = "Get the underlying memref storing the data of a parameter vector.";
+    let arguments = (ins ParameterVectorType:$vector);
+    let results = (outs AnyMemRef:$data);
+    let builders = [
+        OpBuilder<(ins "mlir::Value":$vector), [{
+            ParameterVectorType vectorType = cast<ParameterVectorType>(vector.getType());
+            MemRefType memrefType = MemRefType::get({ShapedType::kDynamic}, vectorType.getElementType());
+            return build($_builder, $_state, memrefType, vector);
+        }]>
+    ];
+    let assemblyFormat = [{ $vector attr-dict `:` type($vector) `->` type($data) }];
+}
```

Comment:
I chose to return a memref because it seemed that this dynamic array type uses mutable semantics that seem more inline with memrefs than with value-based tensor semantics. It could be either, though.

Reply:
Got it, the reason I asked is because we do cast it to a tensor immediately afterwards, but maybe that's just specific to this application here.